hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenClaimAirdropHandler.java

```diff
@@ -36,25 +71,169 @@
  * HederaFunctionality#TOKEN_CLAIM_AIRDROP}.
  */
 @Singleton
-public class TokenClaimAirdropHandler implements TransactionHandler {
+public class TokenClaimAirdropHandler extends TransferExecutor implements TransactionHandler {
+    private final TokenAirdropValidator validator;
 
     @Inject
-    public TokenClaimAirdropHandler() {
-        // Exists for injection
+    public TokenClaimAirdropHandler(
+            @NonNull final TokenAirdropValidator validator,
+            @NonNull final CryptoTransferValidator cryptoTransferValidator) {
+        super(cryptoTransferValidator);
+        this.validator = validator;
     }
 
     @Override
-    public void preHandle(@NonNull PreHandleContext context) throws PreCheckException {}
+    public void preHandle(@NonNull PreHandleContext context) throws PreCheckException {
+        requireNonNull(context);
+        final var op = context.body().tokenClaimAirdrop();
+        requireNonNull(op);
+        final var pendingAirdrops = op.pendingAirdrops();
+
+        final var accountStore = context.createStore(ReadableAccountStore.class);
+
+        for (final var pendingAirdrop : pendingAirdrops) {
+            AccountID receiverId = pendingAirdrop.receiverIdOrThrow();
+            Account account = accountStore.getAccountById(receiverId);
+            requireNonNull(account);
+            Key key = account.key();
+            validateTruePreCheck(key != null, KEY_NOT_PROVIDED);
```

Comment:
We should probably throw `ACCOUNT_IS_IMMUTABLE` similar to other handlers ?

Reply:
I fixed it here. 
Do we want to update crypto transfer validations too? 
Or we can create another issue for this?

Reply:
Since we construct it ourselves, we can add meaningful checks. I have refactored all checks to `validateSemantics` method